Shopify/hydrogen
# 👩‍💻 improved HydrogenSession typing.

### WHY are these changes introduced?

The current typing for `HydrogenSession` does is de-coupled from the implementation of `HydrogenSession` in project. This means the shape of the class could be very different from what @shopify/hydrogen expect session to look like.



### WHAT is this pull request doing?

Export the typing for `HydrogenSession` from @shopify/hydrogen and use it to implement `HydrogenSession` in all the example projects.

The type is implemented using base session type from `@remix-run/server-runtime` which should ensure all session type from remix are supported. (Addressing concern from https://github.com/Shopify/hydrogen/pull/1430#discussion_r1367061880)

Having `HydrogenSession` implemented from a base type will ensure the interface of this class is what we expect it to be when we use property from this class in hydrogen utility. 





### HOW to test your changes?

From any `HydrogenSession` class that implements `DefaultHydrogenSession` type, changed the interface so it no longer conform to the interface and see type error in editor.

example 1: changed return type of `commit` method
![Screenshot 2023-12-20 at 12 44 47 PM](https://github.com/Shopify/hydrogen/assets/1610169/65726e70-e5b0-4e6e-a21a-0cd918b9e210)

example 2: remove `get` method
![Screenshot 2023-12-20 at 12 45 03 PM](https://github.com/Shopify/hydrogen/assets/1610169/848e23dd-cced-4d81-bfcd-360a10d83ddf)




#### Post-merge steps



#### Checklist
